decompiler/frontend/binaryninja/frontend.py

```diff
@@ -79,10 +81,33 @@ def get_all_function_names(self):
             functions.append(function.name)
         return functions
 
-    def _find_function(self, function: str) -> Function:
-        """Return the function at the given address."""
-        address = self._get_address(function)
-        return self._bv.get_function_at(address)
+    def _get_function_from_str(self, symbol_str: str) -> Optional[Function]:
+        """Return Function object matching the given symbol string or hex address"""
+        if symbols_list := self._bv.symbols.get(symbol_str, []):  # alternative: bv.get_symbols_by_name(..)
+            logging.debug(f"symbols list: {symbols_list}")
+            for sym in symbols_list:
+                if sym.type == SymbolType.FunctionSymbol:
+                    if function := self._bv.get_function_at(sym.address):
+                        # check if function name matches symbol. 
+                        # Sometimes wrong symbols are contained in the list.
+                        if function.name == symbol_str:
+                            return function
+            # Sometimes Binja has 2 symbols for a library function, and returns a list:
+            # [ImportedFunctionSymbol, ExternalSymbol]
+            # We want the address of the ImportedFunctionSymbol
+            logging.debug(f"symbols list: {symbols_list}")
+            for sym in symbols_list:
+                if sym.type == SymbolType.ImportedFunctionSymbol:
+                    return self._bv.get_function_at(sym.address)
+        logging.info(f"Did not find matching function for symbol '{symbol_str}'. Try hex address...")
+        try:
+            hex_address = symbol_str[4:] if symbol_str.startswith("sub_") else symbol_str
+            address = int(hex_address, 16)
+            return self._bv.get_function_at(address)
+        except ValueError:
+            logging.info(f"{symbol_str} does not contain a hex value")
+        return None
```

Comment:
Several things here
Frist off, this function is pretty long and has too much nesting. I feel like the symbol iteration got more complex due to reasons we can not fully understand, so it might be a better idea to iterate `bv.functions` instead. I think we could replace the whole top part of the function by something like this:
`next(filter(lambda f: f.name == symbol_str, bv.functions), None)`

Also, it may be worthwhile to extract the second part into its own function as well.
What do you think?

Reply:
I cleaned the symbol stuff up